Python/ceval.c

```diff
@@ -3684,6 +3686,104 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
+            }
+            else {
+                STAT_INC(COMPARE_OP, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                STAT_DEC(COMPARE_OP, unquickened);
+                JUMP_TO_INSTRUCTION(COMPARE_OP);
+            }
+        }
+
+        TARGET(COMPARE_OP_FLOAT) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyFloat_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyFloat_CheckExact(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            double dleft = PyFloat_AS_DOUBLE(left);
+            double dright = PyFloat_AS_DOUBLE(right);
+            int cmp;
+            switch (cache0->original_oparg) {
+                case Py_LT: cmp = (dleft < dright); break;
+                case Py_LE: cmp = (dleft <= dright); break;
+                case Py_EQ: cmp = (dleft == dright); break;
+                case Py_NE: cmp = (dleft != dright); break;
+                case Py_GT: cmp = (dleft > dright); break;
+                case Py_GE: cmp = (dleft >= dright); break;
+                default: Py_UNREACHABLE();
+            }
+            // This cannot fail
+            PyObject *res = PyBool_FromLong(cmp);
+            assert(!PyErr_Occurred());
+            SET_SECOND(res);
+            STACK_SHRINK(1);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            PREDICT(POP_JUMP_IF_FALSE);
+            PREDICT(POP_JUMP_IF_TRUE);
+            DISPATCH();
+        }
+
+        TARGET(COMPARE_OP_INT) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyLong_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyLong_CheckExact(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            // This cannot fail.
+            PyObject *res = _PyLong_RichCompare(
+                (PyLongObject *)left, (PyLongObject *)right,
+                cache0->original_oparg);
+            assert(res != NULL);
+            assert(!PyErr_Occurred());
+            SET_SECOND(res);
+            STACK_SHRINK(1);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            PREDICT(POP_JUMP_IF_FALSE);
+            PREDICT(POP_JUMP_IF_TRUE);
+            DISPATCH();
+        }
+
+        TARGET(COMPARE_OP_STR) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyUnicode_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_CheckExact(right), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
```

Comment:
Is this necessary? If `left == right` then they are equal regardless of whether the string is ready or not.

Reply:
I expect that almost all strings are ready. I suppose the alternative is
```
            DEOPT_IF(!PyUnicode_CheckExact(left), COMPARE_OP);
            DEOPT_IF(!PyUnicode_CheckExact(right), COMPARE_OP);
            int res = 1;
            if (left != right) {
                DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
                DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
                res = comare_the_strings_assuming_theyre_ready(left, right);
            }
```
is that what you mean?

Reply:
Yes. Assuming that the identity shortcut is worthwhile, then it makes sense to shortcut as much work as possible.
You are also missing out of the quick inequality test using lengths `len(a) != len(b)` implies that `a != b`.

Maybe wrap `unicode_compare_eq` in an equality function that readies the strings (I'm surprised that no such thing exists) and use that?

```
int res = 1;
if (left != right) {
    res = _PyUnicode_Equal(left, right);
}